libbeat/cmd/instance/beat.go

```diff
@@ -908,3 +961,23 @@ func initKibanaConfig(beatConfig beatConfig) (*common.Config, error) {
 	}
 	return kibanaConfig, nil
 }
+
+func selectMonitoringConfig(beatCfg beatConfig) (*common.Config, error) {
+	switch {
+	case beatCfg.Monitoring.Enabled() && beatCfg.XPackMonitoring.Enabled():
+		const errMonitoringBothConfigEnabled = "both xpack.monitoring.* and monitoring.* cannot be set. Prefer to set monitoring.* and set monitoring.elasticsearch.hosts to monitoring cluster hosts"
+		return nil, errors.New(errMonitoringBothConfigEnabled)
+	case beatCfg.XPackMonitoring.Enabled():
+		const warnMonitoringDeprecatedConfig = "xpack.monitoring.* settings are deprecated. Use monitoring.* instead, but set monitoring.elasticsearch.hosts to monitoring cluster hosts"
+		cfgwarn.Deprecate("7.0", warnMonitoringDeprecatedConfig)
+		monitoringCfg := beatCfg.XPackMonitoring
+		monitoringCfg.SetInt("_format", -1, int64(report.ReportingFormatXPackMonitoringBulk))
+		return monitoringCfg, nil
+	case beatCfg.Monitoring.Enabled():
+		monitoringCfg := beatCfg.Monitoring
+		monitoringCfg.SetInt("_format", -1, int64(report.ReportingFormatBulk))
+		return monitoringCfg, nil
+	default:
+		return nil, nil
+	}
+}
```

Comment:
Better write a detailed log message, but return a short/descriptive error value. Also think how you error message will print. Most likely: `Failed: both .... cannot be set. Prefer ... cluster host`, not creating a full sentence.

Why do: `const msg = ...;  return nil, errors.New(msg)`? No need to allocate an error. If you have a descriptive name for your error, then declare the error once: `var errMonitoringBothEnabled = errors.New(...)`

btw. can we abstract and move the configuration/loading part from the `instance` package to the `monitoring` (or some other) package?

Reply:
> Better write a detailed log message, but return a short/descriptive error value. Also think how you error message will print. Most likely: Failed: both .... cannot be set. Prefer ... cluster host, not creating a full sentence.

Apologies but I didn't quite understand the problem here. Were you suggesting this `selectMonitoringConfig` function should itself do logging but return a short/descriptive error type? Or something different?

> btw. can we abstract and move the configuration/loading part from the instance package to the monitoring (or some other) package?

I moved the `instance.selectMonitoringConfig` function to `monitoring.SelectConfig`. I had to make a few other changes as well, notably around the `instance.beatConfig` struct. I think it would be good if you could review these changes.

Reply:
Phew... don't remember full context here and I can not find the code pattern or line with recent changes anymore.

Still I see a few `errors.New(<message>)`. This allocates a new error everytime you want to report an error. Errors returned this way can be logged, but are not actionable. One might consider to declare at least a sentinal error constant in the package like `var ErrShortName = errors.New(<common message>)` and reuse the constant.

Reply:
I'm still not 100% sure I got this right but please take a look at the changes I made in 56383a3ae.